File: minswap.js
/*
https://www.hackerrank.com/challenges/minimum-swaps-2/problem?h_l=interview&playlist_slugs%5B%5D=interview-preparation-kit&playlist_slugs%5B%5D=arrays

U
-Unsorted  arr, consecutive INTS [1,2,3,..n], 
-no duplicates 
-Can only swap any 2 elements  
-Min swaps required to sort array in ascending** 
 Ex:  
 arr=[7,1,3,2,4,5,6]

 input= arr Unsorted INTS
  
output = INT, min number of swaps to sort arr  



p
-iterate thru array, 
-if left < right, swap 
  -make sure it covers all 
- track swap  

E   

R


 */

function swapper(arr){
let swap=0
let store={}
let sorted =[]
let res=[]
//subratct arr[i] from i   
//store subtraction in temp arr 

//index of sorted, temp array will give order that swaps take place
//store sub results in object 
//https://www.codecademy.com/forum_questions/50c207bd55df51ff27004775
//go to original array, 
//sort temp arr of objects   
// swap, based temp index order until sorted  
// stop when condition of arr.sort() == arr   
//ed, & swap with its "value"/ true index  
// swap (0,3)
// check if sorted 
//repeat until all sorted 
//res= arr 
console.log(arr)
for(let i=0; i<arr.length; i++){
  let diff= Math.abs(arr[i]-(i+1)) 
  store[i]={
    "diff": diff, "index": i, "value": arr[i]
  }
}

for(var myind in store){
  sorted.push(store[myind])
}

sorted.sort(function(a,b){

return (a.diff - b.diff);

})
sorted.reverse()//sort to descendig 
console.log("sorted", sorted )

//console.log("store", sorted)

//given sorted 
//take index 0 in sorted, & swap with its "value"/ true index  
// swap (0,3)
// check if sorted 
//repeat until all sorted 
//https://stackoverflow.com/questions/4011629/swapping-two-items-in-a-javascript-array

for(let j=0; j<Object.keys(sorted).length; j++){
//console.log("arr3", arr,arr[store[j].diff], arr[j])


function lilswap(arr){
     var tmp = arr[sorted[j].index];
     console.log("here", arr[sorted[j].index],sorted[j].value-1 )
    arr[sorted[j].index]=  arr[sorted[j].value-1];
    arr[sorted[j].value-1] = tmp;
 
  swap+=1
 //console.log("tmp", arr[j],  arr[sorted[j].diff], j)
} 
lilswap(arr)
console.log("swapped", arr)
}

console.log("end", arr,swap)
}

swapper([4,3,1,2])

File: minswap.py

from collections import OrderedDict
from operator import itemgetter

"""
for i in range(0, len(arr)):
    diff= abs(arr[i]-(i+1))
    for x in arr:
      store[i] =diff  
  thank= OrderedDict(sorted(store.items(), key=itemgetter(1)))
  print(thank)
"""

def swapper(arr):
  temp = [0] * (len(arr) +1) #CREATES AN ARR WITH ALL ZEROS, 
  #WITH LEN OF ARR + 1

  for pos, val in enumerate(arr): #creates a key num index, with values for each key
    #print()
    temp[val]= pos #each val in arr is a key of Temp[val], 
    #pos (index) = value of temp arr 
    #print("temp[val]", "val(index)", val,"pos", pos,"tempArr:", temp)
    #pos +=1 # increment index/ arr value in temp arr , #as 
   #temp is created
    swaps = 0
  for i in range(len(arr)):
    if arr[i]!= i+1 :#this condition checks if sorted 
      swaps += 1 #swap increases by 1 
      t = arr[i] # store arr[i](not sorted) in var t! # t=arr[0]=4                     #3=arr[1]
      #print("t0", t)    
      #print("arr[i]", arr[i])      
      arr[i] = i+1 #change value of arr[i] to sorted  # arr[0]= o+1 = 1   
      #print("i:", i, "arr[i] = i+1, arr[i]: ", arr[i],"arr[temp[i+1]]: ", arr[temp[i+1]] )         #arr[1] = 1+1
      arr[temp[i+1]] = t #temp[0+1]=> 3 => arr[3]= t   # arr[temp[0+1] =>3]= t     #arr[temp[1+1]=>2] = t 
      #print("arr[temp[i+1]]: ", arr[temp[i+1]] )
      #print("after arr[temp[i+1]] =t. t= ", t)
      temp[t] = temp[i+1] # temp[3]= temp[0+1] =>  temp[3] = temp[2]  or 1 = 2
      #print("i:", i, "t:", t, "temp[t]:", temp[t], "temp[i+1]:", temp[i+1], "temp:", temp)
  print(swaps)
  return swaps 




#7,1,3,2,4,5,6

#1,3,5,2,4,6,7
  


print(swapper([1,3,5,2,4,6,7]))